// @finos/perspective-viewer/dist/esm/@finos/perspective-vieux/dist/esm/vieux.js
import * as internal from '../../pkg/perspective_vieux.js';
import internal__default, { set_panic_hook } from '../../pkg/perspective_vieux.js';
import wasm_internal from '../../pkg/perspective_vieux_bg.wasm';

function _CustomElement() {
  return Reflect.construct(HTMLElement, [], this.__proto__.constructor);
}
Object.setPrototypeOf(_CustomElement.prototype, HTMLElement.prototype);
Object.setPrototypeOf(_CustomElement, HTMLElement);
const wasm = internal__default(wasm_internal).then(() => {
  set_panic_hook();
  return internal;
});
let _index = undefined;

async function _await_index(f) {
  await new Promise(setTimeout);

  if (!_index) {
    _index = await wasm;
  }

  return f();
} // function handle_view_errors(e) {
//     if (e.message !== "View is not initialized") {
//         throw e;
//     }
// }


class PerspectiveVieuxElement extends _CustomElement {
  constructor() {
    super();

    _await_index(() => {
      this._instance = new _index.PerspectiveVieuxElement(this);
    });
  }

  connectedCallback() {
    _await_index(() => {
      this._instance.connected_callback();
    });
  }

  load(table) {
    _await_index(() => this._instance.load(table));
  }

  set_view(view) {
    _await_index(() => this._instance.set_view(view));
  }

  delete_view() {
    return _await_index(() => this._instance.delete_view());
  }

  toggle_config(force) {
    return _await_index(() => this._instance.toggle_config(force));
  }

  _open_expression_editor(target) {
    return _await_index(() => this._instance._open_expression_editor(target));
  }

}

if (document.createElement("perspective-vieux").constructor === HTMLElement) {
  window.customElements.define("perspective-vieux", PerspectiveVieuxElement);
}

class PerspectiveColumnStyleElement extends _CustomElement {
  constructor() {
    super();
  }

  open(target, config, default_config) {
    _await_index(() => {
      if (this._instance) {
        this._instance.reset(config, default_config);
      } else {
        this._instance = new _index.PerspectiveColumnStyleElement(this, config, default_config);
      }

      this._instance.open(target);
    });
  }

}

if (document.createElement("perspective-column-style").constructor === HTMLElement) {
  window.customElements.define("perspective-column-style", PerspectiveColumnStyleElement);
}

export { wasm };
