// @finos/perspective/dist/cjs/api/view_api.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.view = view;
exports.proxy_view = proxy_view;

var _dispatch = require("./dispatch.js");

/******************************************************************************
 *
 * Copyright (c) 2019, the Perspective Authors.
 *
 * This file is part of the Perspective library, distributed under the terms of
 * the Apache License 2.0.  The full license can be found in the LICENSE file.
 *
 */

/**
 * Construct a proxy for the view object by creating a "view" message and
 * sending it through the worker.
 *
 * @param {*} worker
 * @param {*} table_name
 * @param {*} config
 */
function view(worker, table_name, config) {
  return new Promise((resolve, reject) => {
    var _this$_worker$_featur;

    this._worker = worker;
    this._name = Math.random() + "";

    this._worker.post({
      cmd: "view",
      view_name: this._name,
      table_name: table_name,
      config: config
    }, () => {
      // Resolve the Promise with this function, which is a proxy
      // view that dispatches all view methods to the backend, i.e.
      // a Web Worker or a Python/Node Perspective server. Because
      // arrow functions automatically capture `this` from the
      // surrounding scope, there is no need to explicitly bind the
      // resolving value to `this`.
      resolve(this);
    }, reject);

    if (this._worker._initialized === true && !((_this$_worker$_featur = this._worker._features) !== null && _this$_worker$_featur !== void 0 && _this$_worker$_featur.wait_for_response)) {
      resolve(this);
    }
  });
}
/**
 * Create a reference to a view located on `worker` for use by remote clients.
 *
 * @param {worker} worker the Web Worker at which the view is located.
 * @param {String} name a unique name for the view.
 */


function proxy_view(worker, name) {
  this._worker = worker;
  this._name = name;
}

proxy_view.prototype = view.prototype; // Send view methods that do not create new objects (getters, setters etc.) to
// the queue for processing.

view.prototype.get_config = (0, _dispatch.async_queue)("get_config");
view.prototype.get_min_max = (0, _dispatch.async_queue)("get_min_max");
view.prototype.to_json = (0, _dispatch.async_queue)("to_json");
view.prototype.to_arrow = (0, _dispatch.async_queue)("to_arrow");
view.prototype.to_columns = (0, _dispatch.async_queue)("to_columns");
view.prototype.to_csv = (0, _dispatch.async_queue)("to_csv");
view.prototype.schema = (0, _dispatch.async_queue)("schema");
view.prototype.expression_schema = (0, _dispatch.async_queue)("expression_schema");
view.prototype.column_paths = (0, _dispatch.async_queue)("column_paths");
view.prototype.num_columns = (0, _dispatch.async_queue)("num_columns");
view.prototype.num_rows = (0, _dispatch.async_queue)("num_rows");
view.prototype.set_depth = (0, _dispatch.async_queue)("set_depth");
view.prototype.get_row_expanded = (0, _dispatch.async_queue)("get_row_expanded");
view.prototype.expand = (0, _dispatch.async_queue)("expand");
view.prototype.collapse = (0, _dispatch.async_queue)("collapse");
view.prototype.delete = (0, _dispatch.async_queue)("delete");
view.prototype.col_to_js_typed_array = (0, _dispatch.async_queue)("col_to_js_typed_array");
view.prototype.on_update = (0, _dispatch.subscribe)("on_update", "view_method", true);
view.prototype.remove_update = (0, _dispatch.unsubscribe)("remove_update", "view_method", true);
view.prototype.on_delete = (0, _dispatch.subscribe)("on_delete", "view_method", true);
view.prototype.remove_delete = (0, _dispatch.unsubscribe)("remove_delete", "view_method", true);
